File: Sort/elementary_sort_src/insertion_sort.cpp
#include <array>
#include <iostream>

using namespace std;

template<class RandomIt>
void insertion_sort(RandomIt first, RandomIt last) {
    if (first >= last) {
        return;
    }
    for (RandomIt i = first+1; i != last; ++i) {
        auto val = *i;
        RandomIt j = i-1;
        while (j >= first && *j > val) {
            *(j+1) = *j;
            j--;
        }
        *(j+1) = val;
    }
    return;
}

template<class It>
void printByIt(It begin, It end) {
    for (It curr = begin; curr != end; ++curr) {
        cout << *curr << " ";
    }
    cout << endl;
}

int main() {
    array<int, 10> unorder = {9, 8, 7, 6, 5, 4, 3, 2, 1};
    cout << "before order: ";
    printByIt(unorder.begin(), unorder.end());

    insertion_sort(unorder.begin(), unorder.end());

    cout << "after order: ";
    printByIt(unorder.begin(), unorder.end());
}

File: Sort/elementary_sort_src/insertion_sort.go
package main

import (
	"fmt"
	"sort"
)

func InsertionSort(data sort.Interface) {
	n := data.Len()
	for i := 1; i < n; i++ {
		for j := i; j > 0 && data.Less(j, j-1); j-- {
			data.Swap(j, j-1)
		}
	}
}

func main() {
	unorder := sort.IntSlice{9, 8, 7, 6, 5, 4, 3, 2, 1, 0}
	fmt.Printf("before order: %v\n", unorder)

	InsertionSort(unorder)

	fmt.Printf("after order: %v\n", unorder)
}
